pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -356,44 +360,84 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
+	name, path := "task when expression", "pipelinespec.when"
+	for _, we := range whenExpressions {
+		if err := validatePipelineStringVariable(fmt.Sprintf("input[%s]", we.Input), we.Input, prefix, name, path, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+		for _, val := range we.Values {
+			if err := validatePipelineStringVariable(fmt.Sprintf("values[%s]", val), val, prefix, name, path, paramNames, arrayParamNames); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validateParametersVariablesInParameters(params []Param, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+	name, path := "task parameter", "pipelinespec.params"
+	for _, param := range params {
+		if param.Value.Type == ParamTypeString {
+			if err := validatePipelineStringVariable(fmt.Sprintf("[%s]", param.Name), param.Value.StringVal, prefix, name, path, paramNames, arrayParamNames); err != nil {
+				return err
+			}
+		} else {
+			for _, arrayElement := range param.Value.ArrayVal {
+				if err := validatePipelineArrayVariable(fmt.Sprintf("[%s]", param.Name), arrayElement, prefix, name, path, paramNames, arrayParamNames); err != nil {
 					return err
 				}
-			} else {
-				for _, arrayElement := range param.Value.ArrayVal {
-					if err := validatePipelineVariable(fmt.Sprintf("param[%s]", param.Name), arrayElement, prefix, paramNames); err != nil {
-						return err
-					}
-					if err := validatePipelineArraysIsolated(fmt.Sprintf("param[%s]", param.Name), arrayElement, prefix, arrayParamNames); err != nil {
-						return err
-					}
-				}
 			}
 		}
 	}
 	return nil
 }
 
-func validatePipelineVariable(name, value, prefix string, vars sets.String) *apis.FieldError {
-	return substitution.ValidateVariable(name, value, prefix, "task parameter", "pipelinespec.params", vars)
+func validatePipelineStringVariable(name, value, prefix string, locationName string, path string, stringVars sets.String, arrayVars sets.String) *apis.FieldError {
+	if err := validatePipelineVariable(name, value, prefix, locationName, path, stringVars); err != nil {
+		return err
+	}
+	if err := validatePipelineNoArrayReferenced(name, value, prefix, locationName, path, arrayVars); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validatePipelineArrayVariable(name, value, prefix string, locationName string, path string, stringVars sets.String, arrayVars sets.String) *apis.FieldError {
+	if err := validatePipelineVariable(name, value, prefix, locationName, path, stringVars); err != nil {
+		return err
+	}
+	if err := validatePipelineArraysIsolated(name, value, prefix, locationName, path, arrayVars); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validatePipelineVariable(name, value, prefix string, locationName string, path string, vars sets.String) *apis.FieldError {
```

Comment:
there are 2 "smells" here that i'd like to point out:

1. passing variables through that end up being part of constructing an error <-- this makes me think that it would make sense for there to be a function and/or struct that could know what kind of thing it's validating and control that kind of information vs passing it around
2. there are a lot of layers of functions calling functions. the more layers we add, the more cognitive load we put on folks to understand the code, and also it makes the unit testing story very string: which of these functions should we unit test?

im not sure what the ideal looks like here and if you want to leave it like this its not the end of the world here, but i have a feeling there's a much clearer way to organize this - and one step might be just collapsing some of these functions and tolerating the code repetition

Reply:
replaced it with code repetition, doesn't look good, we might have to refactor pipeline validation soon

Reply:
Pipeline validation refactoring is a long overdue, I would recommend to have it in a separate PR (candidate for `good-first-issue`, nah may be overkill for fresh eyes 🤔 )

Reply:
some of these functions could be moved to `param_types.go`